crates/loam-cli/tests/it/build_clients/contracts.rs

```diff
@@ -0,0 +1,96 @@
+use crate::util::TestEnv;
+use predicates::prelude::*;
+
+#[test]
+fn contracts_built() {
+    let contracts = [
+        "soroban_auth_contract",
+        "soroban_custom_types_contract",
+        "hello_world",
+        "soroban_increment_contract",
+    ];
+    TestEnv::from("soroban-init-boilerplate", |env| {
+        env.set_environments_toml(
+            format!(
+                r#"
+development.accounts = [
+    {{ name = "alice" }},
+]
+
+[development.network]
+rpc-url = "http://localhost:8000/rpc"
+network-passphrase = "Standalone Network ; February 2017"
+
+[development.contracts]
+{}
+"#,
+                contracts
+                    .iter()
+                    .map(|c| format!("{c}.workspace = true"))
+                    .collect::<Vec<String>>()
+                    .join("\n")
+            )
+            .as_str(),
+        );
+
+        let mut contract_stdout_predicates = contracts
+            .iter()
+            .map(|c| {
+                predicates::str::contains(format!("installing \"{c}\" wasm bytecode on-chain"))
+                    .and(predicates::str::contains(format!(
+                        "instantiating \"{c}\" smart contract"
+                    )))
+                    .and(predicates::str::contains(format!(
+                        "binding \"{c}\" contract"
+                    )))
+                    .and(predicates::str::contains(format!(
+                        "importing \"{c}\" contract"
+                    )))
+            })
+            .collect::<Vec<_>>();
+
+        env.loam("build").assert().success().stdout(
+            predicates::str::contains("creating keys for \"alice\"\n")
+                .and(predicates::str::contains(
+                    "using network at http://localhost:8000/rpc\n",
+                ))
+                .and(contract_stdout_predicates.pop().unwrap())
+                .and(contract_stdout_predicates.pop().unwrap())
+                .and(contract_stdout_predicates.pop().unwrap())
+                .and(contract_stdout_predicates.pop().unwrap()),
```

Comment:
@willemneal as you might guess from this code, I would like to use `fold` above, instead of `map`. The first thing to try is writing it like this:

```rs
let mut contract_stdout_predicates = contracts
    .iter()
    .fold(predicates::str::contains(""), |acc, c| {
        acc.and(predicates::str::contains(format!(
            "installing \"{c}\" wasm bytecode on-chain"
        )))
        .and(predicates::str::contains(format!(
            "instantiating \"{c}\" smart contract"
        )))
        .and(predicates::str::contains(format!(
            "binding \"{c}\" contract"
        )))
        .and(predicates::str::contains(format!(
            "importing \"{c}\" contract"
        )))
    })
    .collect::<Vec<_>>();
```

But this gives the following error:

```
mismatched types
expected struct 'ContainsPredicate
found struct
'AndPredicate<AndPredicate<AndPredicate<AndPredicate<ContainsPredicate, ContainsPredicate, str>, ContainsPredicate, str>, ContainsPredicate, str>,
ContainsPredicate, str>' (rustc E0308)
```

So basically, Rust doesn't like that the first run of the loop will have a `ContainsPredicate` but subsequent iterations will have some number of `AndPredicate`s piled on there. But the `.stdout(...)` matcher doesn't care, so how do we tell `fold` that it's ok?

I tried this:

```rs
let mut contract_stdout_predicates = contracts.iter().fold(
    predicates::str::contains("") as IntoOutputPredicate<str>,
    |acc, c| {
        acc.and(predicates::str::contains(format!(
            "installing \"{c}\" wasm bytecode on-chain"
        )))
        .and(predicates::str::contains(format!(
            "instantiating \"{c}\" smart contract"
        )))
        .and(predicates::str::contains(format!(
            "binding \"{c}\" contract"
        )))
        .and(predicates::str::contains(format!(
            "importing \"{c}\" contract"
        )))
    },
);
```

But now I get an even scarier error:

```
 cast to unsized type: `ContainsPredicate` as `dyn IntoOutputPredicate<str>`
```

Any other ideas of things to try?

Reply:
I would first not collect back to a vec but keep it as an iterator. Then you can use it in a for loop.
```rust
let mut res = env.loam("build").assert().success();
for predicate in &contract_stdout_predicates {
    res = res.stdout(predicate);
}
```
Your fighting against the iterator. Let it flow!

Reply:
You might need to make `res` be `&mut Assert`. Or make a new variable with the return type of `.stdout`

Reply:
Or if you wanted to use fold, I would use the Assert. Since `.stdout` can be called each time and not one huge predicate.